File: reporter.py
# Report Generator
# نظام إنشاء التقارير

import json
from datetime import datetime
from colorama import Fore, Style
from tabulate import tabulate

class Reporter:
    """فئة إنشاء التقارير"""
    
    def __init__(self):
        self.results = []
        self.start_time = datetime.now()
    
    def add_result(self, url, param, result):
        """إضافة نتيجة للتقرير"""
        self.results.append({
            'url': url,
            'parameter': param,
            'vulnerable': result.get('vulnerable', False),
            'type': result.get('type', 'N/A'),
            'payload': result.get('payload', 'N/A'),
            'timestamp': datetime.now().isoformat()
        })
    
    def generate_console_report(self):
        """إنشاء تقرير للشاشة"""
        print(f"\n{Fore.YELLOW}{'='*60}{Style.RESET_ALL}")
        print(f"{Fore.YELLOW}         SQL Injection Scanner - Report{Style.RESET_ALL}")
        print(f"{Fore.YELLOW}{'='*60}{Style.RESET_ALL}\n")
        
        vulnerable_count = sum(1 for r in self.results if r['vulnerable'])
        total_count = len(self.results)
        
        print(f"{Fore.CYAN}Summary:{Style.RESET_ALL}")
        print(f"  Total Tests: {total_count}")
        print(f"  {Fore.GREEN}Safe: {total_count - vulnerable_count}{Style.RESET_ALL}")
        print(f"  {Fore.RED}Vulnerable: {vulnerable_count}{Style.RESET_ALL}\n")
        
        if self.results:
            table_data = []
            for result in self.results:
                status = f"{Fore.RED}✗ YES{Style.RESET_ALL}" if result['vulnerable'] else f"{Fore.GREEN}✓ NO{Style.RESET_ALL}"
                table_data.append([
                    result['url'],
                    result['parameter'],
                    status,
                    result['type']
                ])
            
            print(tabulate(table_data, headers=['URL', 'Parameter', 'Vulnerable', 'Type'], tablefmt='grid'))
        
        duration = (datetime.now() - self.start_time).total_seconds()
        print(f"\n{Fore.CYAN}Duration: {duration:.2f} seconds{Style.RESET_ALL}")
    
    def generate_json_report(self, filename='report.json'):
        """حفظ التقرير بصيغة JSON"""
        report = {
            'scan_time': self.start_time.isoformat(),
            'completed_time': datetime.now().isoformat(),
            'total_tests': len(self.results),
            'vulnerable_count': sum(1 for r in self.results if r['vulnerable']),
            'results': self.results
        }
        
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(report, f, indent=2, ensure_ascii=False)
        
        print(f"{Fore.GREEN}[+] تم حفظ التقرير: {filename}{Style.RESET_ALL}")
    
    def generate_html_report(self, filename='report.html'):
        """إنشاء تقرير HTML"""
        vulnerable_count = sum(1 for r in self.results if r['vulnerable'])
        total_count = len(self.results)
        safe_count = total_count - vulnerable_count
        
        html_content = f"""
<!DOCTYPE html>
<html dir="rtl">
<head>
    <meta charset="UTF-8">
    <title>SQL Injection Scanner Report</title>
    <style>
        body {{
            font-family: Arial, sans-serif;
            direction: rtl;
            margin: 20px;
            background-color: #f5f5f5;
        }}
        .header {{
            background-color: #333;
            color: white;
            padding: 20px;
            border-radius: 5px;
            text-align: center;
        }}
        .summary {{
            display: flex;
            gap: 20px;
            margin: 20px 0;
        }}
        .summary-box {{
            flex: 1;
            padding: 15px;
            border-radius: 5px;
            text-align: center;
            color: white;
            font-weight: bold;
        }}
        .safe {{ background-color: #4CAF50; }}
        .vulnerable {{ background-color: #f44336; }}
        .total {{ background-color: #2196F3; }}
        table {{
            width: 100%;
            border-collapse: collapse;
            background-color: white;
            margin-top: 20px;
        }}
        th, td {{
            border: 1px solid #ddd;
            padding: 12px;
            text-align: right;
        }}
        th {{
            background-color: #333;
            color: white;
        }}
        tr:nth-child(even) {{ background-color: #f2f2f2; }}
        .yes {{ color: #f44336; font-weight: bold; }}
        .no {{ color: #4CAF50; font-weight: bold; }}
    </style>
</head>
<body>
    <div class="header">
        <h1>SQL Injection Scanner Report</h1>
        <p>تقرير فحص ثغرات SQL Injection</p>
        <p>الوقت: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}</p>
    </div>
    
    <div class="summary">
        <div class="summary-box total">
            <h3>عدد الاختبارات</h3>
            <h2>{total_count}</h2>
        </div>
        <div class="summary-box safe">
            <h3>آمنة</h3>
            <h2>{safe_count}</h2>
        </div>
        <div class="summary-box vulnerable">
            <h3>معرضة للخطر</h3>
            <h2>{vulnerable_count}</h2>
        </div>
    </div>
    
    <table>
        <tr>
            <th>الرابط</th>
            <th>الحقل</th>
            <th>معرضة للخطر</th>
            <th>نوع الثغرة</th>
            <th>الحمولة</th>
        </tr>
"""
        
        for result in self.results:
            vulnerable_text = '<span class="yes">نعم ✗</span>' if result['vulnerable'] else '<span class="no">لا ✓</span>'
            html_content += f"""
        <tr>
            <td>{result['url']}</td>
            <td>{result['parameter']}</td>
            <td>{vulnerable_text}</td>
            <td>{result['type']}</td>
            <td><code>{result['payload']}</code></td>
        </tr>
"""
        
        html_content += """
    </table>
</body>
</html>
"""
        
        with open(filename, 'w', encoding='utf-8') as f:
            f.write(html_content)
        
        print(f"{Fore.GREEN}[+] تم حفظ التقرير HTML: {filename}{Style.RESET_ALL}")

File: pkg/reporter/reporter.go
package reporter

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/wormsystem-cmd/sql-injection-scanner/pkg/scanner"
)

type Report struct {
	ScanTime        time.Time         `json:"scan_time"`
	URL             string            `json:"url"`
	Vulnerabilities []scanner.ScanResult `json:"vulnerabilities"`
	Statistics      Statistics        `json:"statistics"`
}

type Statistics struct {
	TotalTests    int     `json:"total_tests"`
	Vulnerable    int     `json:"vulnerable"`
	Safe          int     `json:"safe"`
	Duration      float64 `json:"duration"`
}

func GenerateJSONReport(results []scanner.ScanResult, filename string) error {
	report := Report{
		ScanTime:        time.Now(),
		Vulnerabilities: results,
		Statistics: Statistics{
			TotalTests: len(results),
			Vulnerable: len(results),
		},
	}

	data, err := json.MarshalIndent(report, "", "  ")
	if err != nil {
		return err
	}

	return os.WriteFile(filename, data, 0644)
}

func GenerateCSVReport(results []scanner.ScanResult, filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	writer.Write([]string{"URL", "Parameter", "Type", "Payload", "Confidence", "Response Time"})

	for _, result := range results {
		writer.Write([]string{
			result.URL,
			result.Parameter,
			result.Type,
			result.Payload,
			fmt.Sprintf("%.2f", result.Confidence),
			fmt.Sprintf("%dms", result.ResponseTime.Milliseconds()),
		})
	}

	return nil
}

func GenerateHTMLReport(results []scanner.ScanResult, filename string) error {
	htmlContent := `<!DOCTYPE html>
<html dir="rtl">
<head>
    <meta charset="UTF-8">
    <title>SQL Injection Scanner Report</title>
    <style>
        * { margin: 0; padding: 0; box-sizing: border-box; }
        body {
            font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
            direction: rtl;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            min-height: 100vh;
            padding: 20px;
        }
        .container { max-width: 1200px; margin: 0 auto; }
        .header {
            background: white;
            padding: 30px;
            border-radius: 10px;
            box-shadow: 0 10px 30px rgba(0,0,0,0.2);
            margin-bottom: 30px;
            text-align: center;
        }
        .header h1 { color: #333; margin-bottom: 10px; }
        .header p { color: #666; }
        .stats {
            display: grid;
            grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
            gap: 20px;
            margin-bottom: 30px;
        }
        .stat-box {
            background: white;
            padding: 20px;
            border-radius: 10px;
            text-align: center;
            box-shadow: 0 5px 15px rgba(0,0,0,0.1);
        }
        .stat-box h3 { color: #666; margin-bottom: 10px; }
        .stat-box .number { font-size: 32px; font-weight: bold; }
        .vulnerable { color: #e74c3c; }
        .safe { color: #27ae60; }
        .total { color: #3498db; }
        table {
            width: 100%;
            border-collapse: collapse;
            background: white;
            border-radius: 10px;
            overflow: hidden;
            box-shadow: 0 5px 15px rgba(0,0,0,0.1);
        }
        th { background: #667eea; color: white; padding: 15px; text-align: right; }
        td { padding: 12px 15px; border-bottom: 1px solid #eee; }
        tr:hover { background: #f5f5f5; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>🛡️ SQL Injection Scanner Report</h1>
            <p>Advanced SQL Injection Detection Tool | Powered by WORM SYSTEM</p>
            <p style="margin-top: 10px; font-size: 12px; color: #999;">©️ Copyright © 2024 WORM SYSTEM - All Rights Reserved</p>
        </div>
        <div class="stats">
            <div class="stat-box">
                <h3>إجمالي الاختبارات</h3>
                <div class="number total">` + fmt.Sprintf("%d", len(results)) + `</div>
            </div>
            <div class="stat-box">
                <h3>ثغرات مكتشفة</h3>
                <div class="number vulnerable">` + fmt.Sprintf("%d", len(results)) + `</div>
            </div>
        </div>
        <table>
            <thead>
                <tr>
                    <th>الرابط</th>
                    <th>المعامل</th>
                    <th>نوع الثغرة</th>
                    <th>الحمولة</th>
                    <th>درجة الثقة</th>
                </tr>
            </thead>
            <tbody>
`

	for _, result := range results {
		htmlContent += fmt.Sprintf(`
                <tr>
                    <td>%s</td>
                    <td>%s</td>
                    <td>%s</td>
                    <td><code>%s</code></td>
                    <td>%.0f%%</td>
                </tr>
`, result.URL, result.Parameter, result.Type, result.Payload, result.Confidence*100)
	}

	htmlContent += `
            </tbody>
        </table>
    </div>
</body>
</html>
`

	return os.WriteFile(filename, []byte(htmlContent), 0644)
}
